Plugin end-to-end tests do not work out-of-the-box

I started plugin development with Nx and end-to-end tests do not work out-of-the box.

There were two hurdles to jump over:

- Wrong package name
- Nx daemon

I will provide a solution/work-around here. But I think it should be included in the end-to-end test setup or project setup.

Furthermore, I believe the package name is not created as intended or I miss something by reading the [documentation](https://nx.dev/nx-plugin/overview).

## Steps to Reproduce

### Initialize a new Nx plugins repository

We create a new plugin "my-plugin" with a new monorepo "my-org":

```powershell
PS D:\Development\workspaces\git> npx create-nx-plugin@14.0.5 my-org --pluginName my-plugin
Need to install the following packages:
  create-nx-plugin@14.0.5
Ok to proceed? (y) y
Creating a sandbox with Nx...
new my-org "--pluginName" "my-plugin" --preset=empty --collection=@nrwl/workspace
CREATE my-org/nx.json
CREATE my-org/package.json
CREATE my-org/README.md
CREATE my-org/tools/generators/.gitkeep
CREATE my-org/tools/tsconfig.tools.json
CREATE my-org/tsconfig.base.json
CREATE my-org/.editorconfig
CREATE my-org/.gitignore
CREATE my-org/.prettierignore
CREATE my-org/.vscode/extensions.json
CREATE my-org/workspace.json
CREATE my-org/.prettierrc
CREATE my-org/apps/.gitkeep
CREATE my-org/libs/.gitkeep

added 406 packages, and audited 407 packages in 20s

42 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
warning: LF will be replaced by CRLF in .editorconfig.
The file will have its original line endings in your working directory
warning: LF will be replaced by CRLF in .gitignore.
The file will have its original line endings in your working directory
warning: LF will be replaced by CRLF in .prettierignore.
The file will have its original line endings in your working directory
warning: LF will be replaced by CRLF in .prettierrc.
The file will have its original line endings in your working directory
warning: LF will be replaced by CRLF in .vscode/extensions.json.
The file will have its original line endings in your working directory
warning: LF will be replaced by CRLF in README.md.
The file will have its original line endings in your working directory
warning: LF will be replaced by CRLF in nx.json.
The file will have its original line endings in your working directory
warning: LF will be replaced by CRLF in package-lock.json.
The file will have its original line endings in your working directory
warning: LF will be replaced by CRLF in package.json.
The file will have its original line endings in your working directory
warning: LF will be replaced by CRLF in tools/tsconfig.tools.json.
The file will have its original line endings in your working directory
warning: LF will be replaced by CRLF in tsconfig.base.json.
The file will have its original line endings in your working directory
warning: LF will be replaced by CRLF in workspace.json.
The file will have its original line endings in your working directory
Successfully initialized git.

added 4 packages, and audited 411 packages in 5s

42 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
nx generate @nrwl/nx-plugin:plugin my-plugin --importPath=my-org/my-plugin
CREATE packages/my-plugin/package.json
CREATE packages/my-plugin/README.md
CREATE packages/my-plugin/src/index.ts
CREATE packages/my-plugin/tsconfig.json
CREATE packages/my-plugin/tsconfig.lib.json
CREATE packages/my-plugin/project.json
UPDATE workspace.json
UPDATE tsconfig.base.json
CREATE .eslintrc.json
UPDATE .vscode/extensions.json
UPDATE package.json
CREATE packages/my-plugin/.eslintrc.json
CREATE jest.config.ts
CREATE jest.preset.ts
CREATE packages/my-plugin/jest.config.ts
CREATE packages/my-plugin/tsconfig.spec.json
CREATE packages/my-plugin/executors.json
CREATE packages/my-plugin/generators.json
CREATE packages/my-plugin/src/generators/my-plugin/files/src/index.ts__template__
CREATE packages/my-plugin/src/generators/my-plugin/generator.spec.ts
CREATE packages/my-plugin/src/generators/my-plugin/generator.ts
CREATE packages/my-plugin/src/generators/my-plugin/schema.d.ts
CREATE packages/my-plugin/src/generators/my-plugin/schema.json
CREATE packages/my-plugin/src/executors/build/executor.spec.ts
CREATE packages/my-plugin/src/executors/build/executor.ts
CREATE packages/my-plugin/src/executors/build/schema.d.ts
CREATE packages/my-plugin/src/executors/build/schema.json
CREATE e2e/my-plugin-e2e/tests/my-plugin.spec.ts
CREATE e2e/my-plugin-e2e/tsconfig.json
CREATE e2e/my-plugin-e2e/project.json
CREATE e2e/my-plugin-e2e/jest.config.ts
CREATE e2e/my-plugin-e2e/tsconfig.spec.json

added 90 packages, changed 1 package, and audited 501 packages in 8s

61 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
```

That works.

### Invoke end-to-end-tests

Now we want to invoke the end-to-end tests for the just created plugin.

#### First try

```powershell

PS D:\Development\workspaces\git\my-org> npx nx e2e my-plugin-e2e

 FAIL   my-plugin-e2e  e2e/my-plugin-e2e/tests/my-plugin.spec.ts (90.159 s)
  my-plugin e2e
    × should create my-plugin (33383 ms)
    --directory
      × should create src in the specified directory (27338 ms)
    --tags
      × should add tags to the project (28159 ms)

  ● my-plugin e2e › should create my-plugin

    Command failed: npm install

      at runPackageManagerInstall (../../packages/nx-plugin/src/utils/testing-utils/nx-project.ts:54:27)
      at newNxProject (../../packages/nx-plugin/src/utils/testing-utils/nx-project.ts:74:3)
      at ensureNxProject (../../packages/nx-plugin/src/utils/testing-utils/nx-project.ts:86:3)
      at tests/my-plugin.spec.ts:11:20
      at ../../node_modules/tslib/tslib.js:118:75
      at Object.__awaiter (../../node_modules/tslib/tslib.js:114:16)

  ● my-plugin e2e › --directory › should create src in the specified directory

    Command failed: npm install

      at runPackageManagerInstall (../../packages/nx-plugin/src/utils/testing-utils/nx-project.ts:54:27)
      at newNxProject (../../packages/nx-plugin/src/utils/testing-utils/nx-project.ts:74:3)
      at ensureNxProject (../../packages/nx-plugin/src/utils/testing-utils/nx-project.ts:86:3)
      at tests/my-plugin.spec.ts:21:22
      at ../../node_modules/tslib/tslib.js:118:75
      at Object.__awaiter (../../node_modules/tslib/tslib.js:114:16)

  ● my-plugin e2e › --tags › should add tags to the project

    Command failed: npm install

      at runPackageManagerInstall (../../packages/nx-plugin/src/utils/testing-utils/nx-project.ts:54:27)
      at newNxProject (../../packages/nx-plugin/src/utils/testing-utils/nx-project.ts:74:3)
      at ensureNxProject (../../packages/nx-plugin/src/utils/testing-utils/nx-project.ts:86:3)
      at tests/my-plugin.spec.ts:34:22
      at ../../node_modules/tslib/tslib.js:118:75
      at Object.__awaiter (../../node_modules/tslib/tslib.js:114:16)

Test Suites: 1 failed, 1 total
Tests:       3 failed, 3 total
Snapshots:   0 total
Time:        91.523 s
Ran all test suites.

 ————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

 >  NX   Ran target e2e for project my-plugin-e2e (2m)

    ×    1/1 failed
    √    0/1 succeeded [0 read from cache]
```

#### Second try

So `npm install` failed in the end-to-end tests.
Fortunately, we can investigate the created `tmp` directory:

```powershell
PS D:\Development\workspaces\git\my-org\tmp\nx-e2e\proj> npm install
npm ERR! code EINVALIDPACKAGENAME
npm ERR! Invalid package name "my-org/my-plugin": name can only contain URL-friendly characters

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\gnuechtel\AppData\Local\npm-cache\_logs\2022-05-01T11_45_57_526Z-debug.log
```

Now we search 'my-org/my-plugin' in repo's directory and replace it by '@my-org/my-plugin'.

The following files should be changed:

- `tsconfig.base.json`
- `packages/my-plugin/package.json`
- `packages/my-plugin/src/generators/my-plugin/generator.ts`
- `e2e/my-plugin-e2e/tests/my-plugin.spec.ts`

But only one test passes:

```powershell
PS D:\Development\workspaces\git\my-org> npx nx e2e my-plugin-e2e

> nx run my-plugin-e2e:e2e

Compiling TypeScript files for project "my-plugin"...
Done compiling TypeScript files for project "my-plugin".
 FAIL   my-plugin-e2e  e2e/my-plugin-e2e/tests/my-plugin.spec.ts (46.762 s)
  my-plugin e2e
    √ should create my-plugin (45402 ms)
    --directory
      × should create src in the specified directory (2 ms)
    --tags
      × should add tags to the project

  ● my-plugin e2e › --directory › should create src in the specified directory

    EBUSY: resource busy or locked, rmdir 'D:\Development\workspaces\git\my-org/tmp/nx-e2e/proj'

      at rmdirSync (../../node_modules/fs-extra/lib/remove/rimraf.js:264:13)
      at rimrafSync (../../node_modules/fs-extra/lib/remove/rimraf.js:243:7)
      at cleanup (../../packages/nx-plugin/src/utils/testing-utils/utils.ts:125:13)
      at newNxProject (../../packages/nx-plugin/src/utils/testing-utils/nx-project.ts:71:10)
      at ensureNxProject (../../packages/nx-plugin/src/utils/testing-utils/nx-project.ts:86:3)
      at tests/my-plugin.spec.ts:21:22
      at ../../node_modules/tslib/tslib.js:118:75

  ● my-plugin e2e › --tags › should add tags to the project

    EBUSY: resource busy or locked, rmdir 'D:\Development\workspaces\git\my-org/tmp/nx-e2e/proj'

      at rmdirSync (../../node_modules/fs-extra/lib/remove/rimraf.js:264:13)
      at rimrafSync (../../node_modules/fs-extra/lib/remove/rimraf.js:243:7)
      at cleanup (../../packages/nx-plugin/src/utils/testing-utils/utils.ts:125:13)
      at newNxProject (../../packages/nx-plugin/src/utils/testing-utils/nx-project.ts:71:10)
      at ensureNxProject (../../packages/nx-plugin/src/utils/testing-utils/nx-project.ts:86:3)
      at tests/my-plugin.spec.ts:34:22
      at ../../node_modules/tslib/tslib.js:118:75

Test Suites: 1 failed, 1 total
Tests:       2 failed, 1 passed, 3 total
Snapshots:   0 total
Time:        47.529 s, estimated 91 s
Ran all test suites.

 ————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

 >  NX   Ran target e2e for project my-plugin-e2e (52s)

    ×    1/1 failed
    √    0/1 succeeded [0 read from cache]
```

#### Third try (with solution)

So our second try failed, because of locked resources. The most common reason for locked resources are running processes. So our first suspect is Nx daemon, which fortunately reveals itself in task manager.

So the solution is to stop the Nx daemon.

First kill the processes:

![image](https://user-images.githubusercontent.com/3738251/166144911-d96c51d3-cf73-4de9-8c3e-31f89586d8ec.png)

Then set the env variable:

```powershell
$env:NX_DAEMON="false"
```

Now everything works fine:

```powershell
PS D:\Development\workspaces\git\my-org> npx nx e2e my-plugin-e2e

> nx run my-plugin-e2e:e2e

Compiling TypeScript files for project "my-plugin"...
Done compiling TypeScript files for project "my-plugin".
 PASS   my-plugin-e2e  e2e/my-plugin-e2e/tests/my-plugin.spec.ts (116.177 s)
  my-plugin e2e
    √ should create my-plugin (42629 ms)
    --directory
      √ should create src in the specified directory (36259 ms)
    --tags
      √ should add tags to the project (36219 ms)

Test Suites: 1 passed, 1 total
Tests:       3 passed, 3 total
Snapshots:   0 total
Time:        116.933 s
Ran all test suites.
```

Example repository (with fix of package names in the second commit):
https://github.com/gnuechtel/nx-plugin-example

### Environment

```powershell
PS D:\Development\workspaces\git\my-org> npx nx report

 >  NX   Report complete - copy this into the issue template

   Node : 16.13.2
   OS   : win32 x64
   npm  : 8.1.2

   nx : 14.0.5
   @nrwl/angular : Not Found
   @nrwl/cypress : Not Found
   @nrwl/detox : Not Found
   @nrwl/devkit : 14.0.5
   @nrwl/eslint-plugin-nx : 14.0.5
   @nrwl/express : Not Found
   @nrwl/jest : 14.0.5
   @nrwl/js : 14.0.5
   @nrwl/linter : 14.0.5
   @nrwl/nest : Not Found
   @nrwl/next : Not Found
   @nrwl/node : Not Found
   @nrwl/nx-cloud : Not Found
   @nrwl/nx-plugin : 14.0.5
   @nrwl/react : Not Found
   @nrwl/react-native : Not Found
   @nrwl/schematics : Not Found
   @nrwl/storybook : Not Found
   @nrwl/web : Not Found
   @nrwl/workspace : 14.0.5
   typescript : 4.6.4
   rxjs : 6.6.7
   ---------------------------------------
   Community plugins:
```

--- a/e2e/utils/index.ts
+++ b/e2e/utils/index.ts
@@ -173,6 +173,47 @@ export function runCreateWorkspace(
   return create ? create.toString() : '';
 }
 
+export function runCreatePlugin(
+  name: string,
+  {
+    pluginName,
+    packageManager,
+    extraArgs,
+    useDetectedPm = false,
+  }: {
+    pluginName?: string;
+    packageManager?: 'npm' | 'yarn' | 'pnpm';
+    extraArgs?: string;
+    useDetectedPm?: boolean;
+  }
+) {
+  projName = name;
+
+  const pm = getPackageManagerCommand({ packageManager });
+
+  let command = `${pm.runUninstalledPackage} create-nx-plugin ${name}`;
+
+  if (pluginName) {
+    command += ` --pluginName=${pluginName}`;
+  }
+
+  if (packageManager && !useDetectedPm) {
+    command += ` --package-manager=${packageManager}`;
+  }
+
+  if (extraArgs) {
+    command += ` ${extraArgs}`;
+  }
+
+  const create = execSync(command, {
+    cwd: e2eCwd,
+    stdio: [0, 1, 2],
+    env: process.env,
+    encoding: 'utf-8',
+  });
+  return create ? create.toString() : '';
+}
+
 export function packageInstall(
   pkg: string,
   projName?: string,

--- a/e2e/workspace-create/src/create-nx-plugin.test.ts
+++ b/e2e/workspace-create/src/create-nx-plugin.test.ts
@@ -1,0 +1,30 @@
+import {
+  checkFilesExist,
+  getSelectedPackageManager,
+  packageManagerLockFile,
+  runCLI,
+  uniq,
+  runCreatePlugin,
+} from '@nrwl/e2e/utils';
+
+describe('create-nx-plugin', () => {
+  const packageManager = getSelectedPackageManager() || 'pnpm';
+
+  it('should be able to create a plugin repo and run plugin e2e', () => {
+    const wsName = uniq('ws-plugin');
+    const pluginName = uniq('plugin');
+    runCreatePlugin(wsName, {
+      packageManager,
+      pluginName,
+    });
+
+    checkFilesExist(
+      'workspace.json',
+      'package.json',
+      packageManagerLockFile[packageManager],
+      `packages/${pluginName}/package.json`
+    );
+
+    expect(() => runCLI(`e2e ${pluginName}-e2e`)).not.toThrow();
+  });
+});

--- a/packages/create-nx-plugin/bin/create-nx-plugin.ts
+++ b/packages/create-nx-plugin/bin/create-nx-plugin.ts
@@ -98,7 +98,7 @@ function createNxPlugin(
   packageManager,
   parsedArgs: any
 ) {
-  const importPath = parsedArgs.importPath ?? `${workspaceName}/${pluginName}`;
+  const importPath = parsedArgs.importPath ?? `@${workspaceName}/${pluginName}`;
   const command = `nx generate @nrwl/nx-plugin:plugin ${pluginName} --importPath=${importPath}`;
   console.log(command);
 

--- a/packages/nx-plugin/generators.ts
+++ b/packages/nx-plugin/generators.ts
@@ -1,0 +1,5 @@
+export * from './src/generators/e2e-project/e2e';
+export * from './src/generators/executor/executor';
+export * from './src/generators/generator/generator';
+export * from './src/generators/plugin/plugin';
+export * from './src/generators/migration/migration';

--- a/packages/nx-plugin/src/generators/e2e-project/files/tests/__pluginName__.spec.ts__tmpl__
+++ b/packages/nx-plugin/src/generators/e2e-project/files/tests/__pluginName__.spec.ts__tmpl__
@@ -5,36 +5,54 @@ import {
   runNxCommandAsync,
   uniq,
 } from '@nrwl/nx-plugin/testing';
+
 describe('<%= pluginName %> e2e', () => {
-    it('should create <%= pluginName %>', async () => {
-        const plugin = uniq('<%= pluginName %>');
-        ensureNxProject('<%= npmPackageName %>', '<%= pluginOutputPath %>');
-        await runNxCommandAsync(`generate <%=npmPackageName%>:<%= pluginName %> ${plugin}`);
+  // Setting up individual workspaces per
+  // test can cause e2e runs to take a long time.
+  // For this reason, we recommend each suite only
+  // consumes 1 workspace. The tests should each operate
+  // on a unique project in the workspace, such that they
+  // are not dependant on one another.
+  beforeAll(() => {
+    ensureNxProject('<%= npmPackageName %>', '<%= pluginOutputPath %>');
+  });
 
-        const result = await runNxCommandAsync(`build ${plugin}`);
-        expect(result.stdout).toContain('Executor ran');
-    }, 120000)
+  afterAll(() => {
+    // `nx reset` kills the daemon, and performs
+    // some work which can help clean up e2e leftovers
+    runNxCommandAsync('reset');
+  });
 
-    describe('--directory', () => {
-      it('should create src in the specified directory', async () => {
-        const plugin = uniq('<%= pluginName %>');
-        ensureNxProject('<%= npmPackageName %>', '<%= pluginOutputPath %>');
-        await runNxCommandAsync(
-          `generate <%=npmPackageName%>:<%= pluginName %> ${plugin} --directory subdir`
-        );
-        expect(() => checkFilesExist(`libs/subdir/${plugin}/src/index.ts`)).not.toThrow();
-      }, 120000);
-    });
+  it('should create <%= pluginName %>', async () => {
+    const project = uniq('<%= pluginName %>');
+    await runNxCommandAsync(
+      `generate <%=npmPackageName%>:<%= pluginName %> ${project}`
+    );
+    const result = await runNxCommandAsync(`build ${project}`);
+    expect(result.stdout).toContain('Executor ran');
+  }, 120000);
 
-    describe('--tags', () => {
-      it('should add tags to the project', async () => {
-         const plugin = uniq('<%= pluginName %>');
-        ensureNxProject('<%= npmPackageName %>', '<%= pluginOutputPath %>');
-        await runNxCommandAsync(
-          `generate <%=npmPackageName%>:<%= pluginName %> ${plugin} --tags e2etag,e2ePackage`
-        );
-        const project = readJson(`libs/${plugin}/project.json`);
-        expect(project.tags).toEqual(['e2etag', 'e2ePackage']);
-      }, 120000);
-    });
-})
+  describe('--directory', () => {
+    it('should create src in the specified directory', async () => {
+      const project = uniq('<%= pluginName %>');
+      await runNxCommandAsync(
+        `generate <%=npmPackageName%>:<%= pluginName %> ${project} --directory subdir`
+      );
+      expect(() =>
+        checkFilesExist(`libs/subdir/${project}/src/index.ts`)
+      ).not.toThrow();
+    }, 120000);
+  });
+
+  describe('--tags', () => {
+    it('should add tags to the project', async () => {
+      const projectName = uniq('<%= pluginName %>');
+      ensureNxProject('<%= npmPackageName %>', '<%= pluginOutputPath %>');
+      await runNxCommandAsync(
+        `generate <%=npmPackageName%>:<%= pluginName %> ${projectName} --tags e2etag,e2ePackage`
+      );
+      const project = readJson(`libs/${projectName}/project.json`);
+      expect(project.tags).toEqual(['e2etag', 'e2ePackage']);
+    }, 120000);
+  });
+});

